Ubuntu/History/historymanager.cpp

```diff
@@ -0,0 +1,186 @@
+{
+    return mFilter;
+}
+
+void HistoryManager::setFilter(HistoryQmlFilter *value)
+{
+    if (mFilter) {
+        mFilter->disconnect(this);
+    }
+
+    mFilter = value;
+
+    Q_EMIT filterChanged();
+}
+
+HistoryModel::EventType HistoryManager::type() const
+{
+    return mType;
+}
+
+void HistoryManager::setType(HistoryModel::EventType value)
+{
+    mType = value;
+    Q_EMIT typeChanged();
+}
+
+HistoryQmlSort *HistoryManager::sort() const
+{
+    return mSort;
+}
+
+int HistoryManager::count() const
+{
+    return mCount;
+}
+
+int HistoryManager::deletedCount() const
+{
+    return mDeletedCount;
+}
+
+void HistoryManager::setSort(HistoryQmlSort *value)
+{
+    // disconnect the previous sort
+    if (mSort) {
+        mSort->disconnect(this);
+    }
+
+    mSort = value;
+
+    Q_EMIT sortChanged();
+}
+
+int HistoryManager::eventsCount()
+{
+    History::Filter queryFilter;
+
+    if (!mType || !mFilter) {
+        return false;
+    }
+
+    if (mFilter && mFilter->filter().isValid()) {
+        queryFilter = mFilter->filter();
+    }
+
+    return History::Manager::instance()->eventsCount((History::EventType)mType, queryFilter);
+}
+
+void HistoryManager::timerEvent(QTimerEvent *event)
+{
+    if (event->timerId() == mSignalsTimer) {
+        pendingOperation(false);
+    }
+}
+
+bool HistoryManager::removeAll()
+{
+
+    History::Filter queryFilter;
+    History::Sort querySort;
+
+    if (mPendingOperation) {
+        qWarning() << "there is a pending operation, request cancelled";
+        return false;
+    }
+
+    if (!mType || !mFilter || !mFilter->filter().isValid()) {
+        qWarning() << "type or filter not set, operation cancelled";
+        return false;
+    }
+
+    if (!mView.isNull()) {
+        mView->disconnect(this);
+    }
+
+    queryFilter = mFilter->filter();
+
+    if (mSort) {
+        querySort = mSort->sort();
+    }
+
+    setCount(eventsCount());
+    mDeletedCount = 0;
+
+    bool started = false;
+    if (count() > 0) {
+
+        mView = History::Manager::instance()->queryEvents((History::EventType)mType, querySort, queryFilter);
+        connect(mView.data(),
+                SIGNAL(eventsRemoved(History::Events)),
+                SLOT(onEventsRemoved(History::Events)));
+
+        started = History::Manager::instance()->removeEvents((History::EventType)mType, queryFilter, querySort);
+
+    }
+    pendingOperation(started);
+
+    return started;
+}
+
+void HistoryManager::onEventsRemoved(const History::Events &events)
+{
+    mDeletedCount += events.count();
+    qDebug() << "mDeletedCount" << mDeletedCount;
+    Q_EMIT deletedCountChanged();
+
+    if (mDeletedCount == mCount) {
+        pendingOperation(false);
+    }
+}
+
+void HistoryManager::pendingOperation(bool state)
+{
+    mPendingOperation = state;
+    if (mPendingOperation) {
+        if (mSignalsTimer >= 0) {
+            killTimer(mSignalsTimer);
+        }
+
+        mSignalsTimer = startTimer(180000);
```

Comment:
Am I reading this right, that a pending operation is just dropped after 180 seconds? Where is that timeout handled? What if it leaves us in an inconsistent state, or a client starts another operation not knowing that the last one is running?

You might also consider using [`QTimer`](https://doc.qt.io/qt-5/qtimer.html) instead of `QObject::startTimer()` so you don't have to fiddle with timer IDs and can connect signals and slots.

Reply:
Yes that is a better-than-nothing solution and just did it to make sure app will not get infinite loading.
Deleting 10000 contact history calls take about a minute on BQ4.5.
There is no inconsistency or concurrency issue afaik, because operation#2 will not start until operation#1 is finished ( at least my different tests concluded that).
 
App is notified that operation timed out with the signal `operationEnded` and by knowing if there are remaining events to delete.